content/2022-02-25-pattern-matching-with-named-fields.md

```diff
@@ -0,0 +1,306 @@
+---
+
+## History
+
+| Date          | Version       |
+|---------------|---------------|
+| Feb 25th 2022 | Initial Draft |
+
+## Motivation
+
+An readable, extensible, and intuitive way to deconstruct case classes in pattern matching.
+
+Link to work in progress implementation lives here: https://github.com/Jentsch/dotty
+
+## Motivating Examples
+
+Given one wants to use pattern matching on a case class:
+
+```scala
+case class User(name: String, age: Int, city: String)
+
+val user = User(name = "Anna", age = 10, city = "Berlin")
+
+val annasCity = user match
+  case User(name = "Anna", city = c) => c
+```
+
+The Deconstruction allows the same syntax as the construction and seems to be what people intuitively expect. See for example the very first post in [Scala Contributors Thread][contributors-thread] for this topic.
+
+Without names in patterns users have to use underscore a lot. The example above would be written as:
+
+//TODO: how offend does this pattern occur?
+
+```scala
+val annasCity = user match
+  case User("Anna", _, c) => c
+```
+
+This makes it hard which parameter means what, basically the same idea as for named arguments. Adding underscores until the compile is happy is also not a great experience. IDEs help here, by showing the names.
+
+In addition, the code breaks every time a field of `User` gets added, rearranged, or removed.
+In the worst case it breaks silently, if two fields with the same type switch places.
+
+My personal motivation comes from using [`matchPattern` in ScalaTest](https://www.scalatest.org/user_guide/using_matchers#matchingAPattern)
+and got bitten by it every time my data model changed slightly.
+
+## Counter-Examples
+
+This SIP doesn't aim to allow pattern where parameters go into the pattern, e.g:
+
+```scala
+val map = Map("Berlin" -> 10, "Paris" -> 5)
+
+map match
+  case Map("Paris" = five) => five
+```
+
+## Design
+
+THe goal is to allow named parameter in deconstruction as in construction.
+
+### Mixed usage
+
+Mixed patterns, with positional and named patterns are allowed to keep the similarity.
+But they have no motivational use case. Maybe they should be disallowed.
+
+```scala
+  case User("Anna", city = c) => // Mixed usage seems wired
+  case User(_, city = c) => // Leading underscore are especially useless
+```
+
+### Disallow same name twice
+
+```scala
+  case User(city = city1, city = city2) => a // error city is used twice
+  case User(name1, name = name2) => a // error city is used twice
+```
+
+The same should happen if the `User` had a field with a [deprecated name](https://www.scala-lang.org/api/current/scala/deprecatedName.html).
+
+### Order of name and term
+
+Normally an equal sign assigns the result of the right side to the left side. With the proposed syntax that's not the case. However, for most, if not all, people that's the correct order.
+
+This order seems the requires a single look ahead in the parser.
+
+### Desugaring
```

Comment:
It's not clear what happens for extractors with `unapplySeq` methods (and case classes with variadic arguments).

Reply:
With the given encoding I see no meaningful names that could be given to variadic argument. (Only `first`, `second`, `third` ... would be possible, but not `last`).
I added this constrain down below to the [Con list, line 113](https://github.com/scala/improvement-proposals/pull/44/files#diff-c3cfd3b89047c3905b5786ac634ef23cbab2b40c4f5b2e3913f17342fb5afebcR235)